### ISSUE TITLE
Edge case in both miter/bevel linejoin modes

### ISSUE THREAD

**Issue by kuraiookami** [love2d/love#1894]
This is such a minor bug that it feel pointless to report, but it could confuse/upset someone new to love2d/coding and that would be unfortunate.

For fun I'm working on a love2d controller of the derail valley dispatcher mod. While working on rendering the tracks I ended up with a set of points for a track segment that when rendered cause what should be a short section of track to be a line that runs off screen in both directions. When you set the linejoin mode to bevel one side corrects but the other does not. If you set linejoin mode to none then it renders fine. You can also reduce the decimal precision down to remove the errors but the points listed are generated dynamically while zooming/panning the map.

I have extracted the points and they still reliably glitched for me on both my Linux and Windows machines. I should add that this error occurs when the map is fully zoomed out, that is why this line segment is so short. This is on version 11.3.

function love.draw()
	love.graphics.setLineJoin( "miter" )
	--love.graphics.setLineJoin( "bevel" )
	--love.graphics.setLineJoin( "none" )
	love.graphics.line({ 121.852846370545,109.87029686964,
		121.672390327231,109.35841847852,
		121.491941615924,108.84652359026,
		121.308234826955,108.32542462031,
		121.311489571886,108.33464519913 })
end

Screenshots of the errors with the different linejoin modes
![errors](https://user-images.githubusercontent.com/16405605/212218913-14ca081d-a83a-4dd0-8000-9155ec84eb33.png)

Zipped version of the code above
[main.zip](https://github.com/love2d/love/files/10407643/main.zip)

**Comment by steveRoll-git** [comment]
Out of curiosity I tried zooming in on it, this is what it looks like (with the points themselves shown in red):
![image](https://user-images.githubusercontent.com/40004112/212492936-7f3e98d4-04ac-45f7-87a1-e5936f81bfe3.png)

**Comment by slime73** [comment]
I can reproduce the bug with just the last 3 points in the line, which might help I suppose.

I did a little debugging in the code for the bevel mode (after making some variable names more readable). The problem manifests in the  [renderEdge function](https://github.com/love2d/love/blob/8cac37b811ac8b8292af0cdba7b42b248f248410/src/modules/graphics/Polyline.cpp#L228) - the `det` value is a very small number, and `newSegmentLength` is also very small so the "lines parallel" branch doesn't get hit. `lambda` ends up being really big because `det` is so small, and that big number is used to create some of the vertices of the line.

I don't understand the math behind this stuff well enough to know what to change, though.

The miter mode probably has a similar problem and also likely runs into another issue where it doesn't fall back to something like bevel when the angle between points is small or the length of the miter join is big (which is a common thing to do in line rendering I think).

**Comment by scurest** [comment]
I looked at miter (but not bevel). The bug happens when the two segments are parallel but in opposite directions, ie. the line doubles back on itself. Here's a simpler example

```lua
   -- this draws nothing (det=0, normal=infinity)
    love.graphics.line({
        0, 0,
        100, 100,
        50, 50
    })
    -- this draws a huge line like in the OP
    love.graphics.line({
        0, 0,
        100, 100,
        50, 50.0001
    })
```

This is because the "is parallel" branch also requires the segments be going the same way (dot product > 0)

https://github.com/love2d/love/blob/8cac37b811ac8b8292af0cdba7b42b248f248410/src/modules/graphics/Polyline.cpp#L191

If the dot product condition is removed, both OP's example and the above example work.

I'm not sure what that condition is for. It was added in [`8b53bc3`: Fix line drawing artefact with nearly parallel segments](https://github.com/love2d/love/commit/8b53bc385e2342dcd561563b329754597d40a1bd). It's been nearly a decade but maybe @vrld can remember?

(Btw the new variable names make this code waaaaay harder to understand. The old names, p, q, r, etc. were explained perfectly by the comments.)

**Comment by slime73** [comment]
I do think there are probably two separate issues with miter line joins (no fallback to something like bevel at tight angles, plus the same issue as bevel joins), whereas bevel probably just has one since it shouldn't need any fallback, so IMO it's best to figure out bevel first.

> Btw the new variable names make this code waaaaay harder to understand. The old names, p, q, r, etc. were explained perfectly by the comments.

The reason bugs like these have existed for a whole decade is because no one really understood the code. With the old variables you had to constantly cross-reference with the comment diagrams to understand what they are - and the letters are similar enough that it's extremely easy to get them mixed up. I'm sure the code still isn't perfect, but keeping it as it was isn't a good option for future maintainability, and to the people who maintain the code (e.g. me) the new variables are much more understandable than the old ones. I can see how that might be different for people coming from a math background though, but most of us don't.

**Comment by scurest** [comment]
Bevel has the same issue: the same dot product test, from the same commit.

For bevel the limiting position of the u points really is the same for both parallel and antiparallel segments, so again I'm not sure what the dot product is for.

![out](https://user-images.githubusercontent.com/11024420/220427171-44646f27-a3d7-4a64-969a-b558b3b07084.gif)

**Comment by vrld** [comment]
@scurest I don't remember, but I think the dot product test catches a case where the angle between segments becomes really small so that the intersection of the outer segment edges will be far away from the edge. Rough sketch:
![image](https://user-images.githubusercontent.com/156469/220436635-1327b849-eb31-4eb5-b311-e9a053aa1f72.png)

This makes only sense for the miter join though, for the bevel join it should make no difference as you stated.

**Comment by scurest** [comment]
@vrld But the angle being really small would be caught by the check for the cross product being small, and the dot product test excludes that case since s and t are pointing opposite directions.

**Comment by scurest** [comment]
Another issue is stationary points on the line break it

```lua

function love.draw(dt)
    love.graphics.setLineWidth(20)
    love.graphics.setLineJoin("bevel")
    love.graphics.line({
        0, 200,
        200, 200,
        200, 200,  -- comment this line to fix it
        400, 200
    })
end
```

**Comment by slime73** [comment]
Thanks for the thorough investigation. I've applied your fix to 12.0 - I might consider backporting it to the main/11.x branch too, if we end up wanting to have a release of that before 12.0 comes out.
